### ISSUE TITLE
opscode-push-jobs-client_1.1.5-1_amd64.deb installing Chef 11.x

### ISSUE THREAD

**Issue by scalp42** [chef/opscode-pushy-client#59]
Hey folks,

I'm not sure if this is the right repo, but I'm hitting a blocker here.

I installed `opscode-push-jobs-client_1.1.5-1_amd64.deb` with the push-jobs cookbook and noticed that it was breaking.

A quick look at `push-jobs-client.rb`, I saw that it was pulling the `/etc/chef/client.rb`.

I currently use the Hipchat gem as an exception handler to output a notification, so I noticed the error in the log:

```
2015-06-19_22:11:36.10447 [2015-06-19T22:11:36+00:00] WARN: Failed to load hipchat/chef. This should be resolved after a chef run.
2015-06-19_22:11:36.10462 [2015-06-19T22:11:36+00:00] FATAL: Configuration error NameError: uninitialized constant Class::HipChat
2015-06-19_22:11:36.10469 [2015-06-19T22:11:36+00:00] FATAL:   /etc/chef/push-jobs-client.rb:7:in `from_string'
```

I tried to install the `hipchat` gem in the `opscode-push-jobs-client` omnibus business:

```
root@jumphost-i-5ea5f3a9 [gz-infra] /opt/opscode-push-jobs-client/embedded/bin # ./gem install hipchat
Successfully installed hipchat-1.5.1
1 gem installed
```

and tweaked `/etc/chef/push-jobs-client.rb` to require the gem to get around the `NameError`:

```ruby
require 'hipchat/chef'

Chef::Config.from_file("/etc/chef/client.rb")

# The whitelist comes from node['push_jobs']['whitelist']
whitelist({"chef-client"=>"chef-client", "who"=>"w"})

# We're under runit, so don't output timestamp
Mixlib::Log::Formatter.show_time = false
```

Restarting the `opscode-push-jobs-client` service, I noticed the other issue regarding the `json` gem:

```
2015-06-19_22:16:04.59211 [2015-06-19T22:16:04+00:00] FATAL: Configuration error Gem::LoadError: Unable to activate httparty-0.13.5, because json-1.5.5 conflicts with json (~> 1.8)
2015-06-19_22:16:04.59229 [2015-06-19T22:16:04+00:00] FATAL:   /etc/chef/push-jobs-client.rb:5:in `from_string'
2015-06-19_22:16:04.59237 [2015-06-19T22:16:04+00:00] FATAL: Aborting due to error in '/etc/chef/push-jobs-client.rb'
```

I remember seeing this issue and a quick look at the gems installed by the package gave me the problem:

```
root@jumphost-i-5ea5f3a9 [gz-infra] /opt/opscode-push-jobs-client/embedded/bin # ./gem  list | grep chef
chef (11.12.8)
chef-zero (2.2, 2.0.2)
```

Chef 11 is still using an old version of the json gem.

Am I missing something ? Is there a more updated package version that bundles Chef 12.x version?

Thanks for looking into it!

**Comment by scalp42** [comment]
Went a little further by rolling back to an ancient `httparty` gem version that doesnt lock the `json` gem (v0.11.0). Still no dice:

```
2015-06-19_22:32:10.29331 [2015-06-19T22:32:10+00:00] FATAL: Configuration error NameError: uninitialized constant Chef::Handler
2015-06-19_22:32:10.29339 [2015-06-19T22:32:10+00:00] FATAL:   /etc/chef/push-jobs-client.rb:6:in `from_string'
2015-06-19_22:32:10.29342 [2015-06-19T22:32:10+00:00] FATAL: Aborting due to error in '/etc/chef/push-jobs-client.rb'
```

**Comment by scalp42** [comment]
Looks like dropping a `require 'chef'` in `/etc/chef/push-jobs-client.rb` does the trick.

Running `/opt/opscode-push-jobs-client/embedded/bin/ruby /opt/opscode-push-jobs-client/bin/pushy-client -c /etc/chef/push-jobs-client.rb`:

```
INFO: [jumphost-i-5ea5f3a9] Using node name: jumphost-i-5ea5f3a9
INFO: [jumphost-i-5ea5f3a9] Using Chef server: blabla.com/blabla
INFO: [jumphost-i-5ea5f3a9] Using private key: /etc/chef/client.pem
INFO: [jumphost-i-5ea5f3a9] Using org name: blabla
INFO: [jumphost-i-5ea5f3a9] Incarnation ID: bb2c1f2d-45ed-4913-bc99-1016c4ffac7c
INFO: [jumphost-i-5ea5f3a9] Starting client ...
INFO: [jumphost-i-5ea5f3a9] Retrieving configuration from blabla.com/jumphost-i-5ea5f3a9 ...
INFO: [jumphost-i-5ea5f3a9] Connecting to command channel at tcp://blabla.com:10002
INFO: [jumphost-i-5ea5f3a9] Listening for server heartbeat at tcp://blabla.com:10000
INFO: [jumphost-i-5ea5f3a9] Starting heartbeat / offline detection thread on interval 1.0 ...
INFO: [jumphost-i-5ea5f3a9] Starting command / server heartbeat receive thread ...
INFO: [jumphost-i-5ea5f3a9] Started client.
INFO: [jumphost-i-5ea5f3a9] Starting reconfigure thread.  Will reconfigure / reload keys after 3600 seconds.
INFO: [jumphost-i-5ea5f3a9] Received abort
```

**Comment by scalp42** [comment]
Even if I change the template for `/etc/chef/push-jobs-client.rb` to add that extra `require 'chef'`, I'm confused why it's needed.

**Comment by scalp42** [comment]
If anyone comes across the same issue, here is how I fixed it:

```ruby
include_recipe 'push-jobs'

chef_gem 'chef-rewind'
require 'chef/rewind'

rewind %|template[#{PushJobsHelper.config_path}]| do
  source 'my-push-jobs-client.rb.erb'
  cookbook_name "gz-common"
end

# FIXME: https://github.com/chef/opscode-pushy-client/issues/59
gem_package 'httparty' do
  version '0.11.0'
  gem_binary '/opt/opscode-push-jobs-client/embedded/bin/gem'
  action :install
end

gem_package 'hipchat' do
  gem_binary '/opt/opscode-push-jobs-client/embedded/bin/gem'
  action :install
end
```

`my-push-jobs-client.rb.erb' template:

```
# Generated by Chef for <%= node['fqdn'] %>
#
# Local modifications will be overwritten!

require 'chef'

Chef::Config.from_file("/etc/chef/client.rb")

# The whitelist comes from node['push_jobs']['whitelist']
whitelist(<%= @whitelist.to_hash.inspect %>)

# We're under runit, so don't output timestamp
Mixlib::Log::Formatter.show_time = false
```

**Comment by scalp42** [comment]
Fixed with versions > 1.15.x
